src/core/server/logging/logging_service.ts

```diff
@@ -16,112 +16,88 @@
+   *
+   * @example
+   * Customize the configuration for the plugins.data.search context.
+   * ```ts
+   * core.logging.configure(
+   *   of({
+   *     appenders: new Map(),
+   *     loggers: [{ context: 'search', appenders: ['default'] }]
+   *   })
+   * )
+   * ```
+   *
+   * @param config$
    */
-  public upgrade(rawConfig: LoggingConfigType) {
-    const config = new LoggingConfig(rawConfig);
-    // Config update is asynchronous and may require some time to complete, so we should invalidate
-    // config so that new loggers will be using BufferAppender until newly configured appenders are ready.
-    this.config = undefined;
-
-    // Appenders must be reset, so we first dispose of the current ones, then
-    // build up a new set of appenders.
-    for (const appender of this.appenders.values()) {
-      appender.dispose();
-    }
-    this.appenders.clear();
+  configure(config$: Observable<LoggerContextConfigInput>): void;
+}
 
-    for (const [appenderKey, appenderConfig] of config.appenders) {
-      this.appenders.set(appenderKey, Appenders.create(appenderConfig));
-    }
+/** @internal */
+export interface InternalLoggingServiceSetup {
+  configure(contextParts: string[], config$: Observable<LoggerContextConfigInput>): void;
+}
 
-    for (const [loggerKey, loggerAdapter] of this.loggers) {
-      loggerAdapter.updateLogger(this.createLogger(loggerKey, config));
-    }
+interface SetupDeps {
+  loggingSystem: ILoggingSystem;
+}
 
-    this.config = config;
+/** @internal */
+export class LoggingService implements CoreService<InternalLoggingServiceSetup> {
+  private readonly subscriptions = new Map<string, Subscription>();
+  private readonly log: Logger;
 
-    // Re-log all buffered log records with newly configured appenders.
-    for (const logRecord of this.bufferAppender.flush()) {
-      this.get(logRecord.context).log(logRecord);
-    }
+  constructor(coreContext: CoreContext) {
+    this.log = coreContext.logger.get('logging');
   }
 
-  /**
-   * Disposes all loggers (closes log files, clears buffers etc.). Service is not usable after
-   * calling of this method until new config is provided via `upgrade` method.
-   * @returns Promise that is resolved once all loggers are successfully disposed.
-   */
-  public async stop() {
-    for (const appender of this.appenders.values()) {
-      await appender.dispose();
-    }
-
-    await this.bufferAppender.dispose();
-
-    this.appenders.clear();
-    this.loggers.clear();
+  public setup({ loggingSystem }: SetupDeps) {
+    return {
+      configure: (contextParts: string[], config$: Observable<LoggerContextConfigInput>) => {
+        const contextName = LoggingConfig.getLoggerContext(contextParts);
+        this.log.debug(`Setting custom config for context [${contextName}]`);
+
+        const existingSubscription = this.subscriptions.get(contextName);
+        if (existingSubscription) {
+          existingSubscription.unsubscribe();
+        }
```

Comment:
That's correct. I did it this way to simplify the implementation and avoid having to implement any merging logic. For plugin usages, this should be totally fine since each plugin has it's own logging context.